internal/pkg/image/driver/gocryptfsdriver.go

```diff
@@ -0,0 +1,464 @@
+	sylog.Debugf("cmd starts, cmd prefix: %s, args: %s", cmdPrefix, args)
+	if args[0], err = exec.LookPath(args[0]); err == nil {
+		var cmdArgs []string
+		cmdArgs = append(cmdArgs, cmdPrefix...)
+		cmdArgs = append(cmdArgs, args...)
+		cmd := exec.Command(cmdArgs[0], cmdArgs[1:]...)
+		if stdin != nil {
+			cmd.Stdin = stdin
+		}
+		if sysAttr != nil {
+			cmd.SysProcAttr = sysAttr
+		}
+		if len(extraFiles) > 0 {
+			cmd.ExtraFiles = extraFiles
+		}
+		if err := cmd.Start(); err != nil {
+			return cmd, err
+		}
+		return cmd, nil
+	}
+	return nil, fmt.Errorf("%s could not be located", args[0])
+}
+
+func (g *gocryptfsFeature) stop(target string, kill bool) error {
+	// not mounted using squashfuse or gocryptfs
+	if len(g.mountInfos) == 0 {
+		return nil
+	}
+
+	// normal case
+	if len(g.mountInfos) == 3 {
+		// not our target
+		if g.mountInfos[2].dstPath != target {
+			return nil
+		}
+	}
+
+	sylog.Debugf("gocryptfsFeature stop is called, target: %s, kill: %t, mountInfos: %v", target, kill, g.mountInfos)
+	defer g.cleanFn()
+	// we need to unmount mount points by reverse order
+	for i := len(g.mountInfos) - 1; i >= 0; i-- {
+		defaultCleanFnGenerator(g.mountInfos[i].dstPath, g.mountInfos[i].pid, g.mountInfos[i].mountType)(kill)
+	}
+
+	return nil
+}
+
+func (d *gocryptfsDriver) Features() image.DriverFeature {
+	var features image.DriverFeature
+	if d.gocryptfsFeature.gocryptfsPath != "" && d.gocryptfsFeature.squashfusePath != "" {
+		features |= image.ImageFeature
+	}
+
+	return features
+}
+
+func (d *gocryptfsDriver) Mount(params *image.MountParams, mfunc image.MountFunc) error {
+	if params.Filesystem != "gocryptfs" {
+		return fmt.Errorf("filesystem type should be 'gocryptfs'")
+	}
+	g := &d.gocryptfsFeature
+	return g.Mount(params, mfunc, d.cmdPrefix)
+}
+
+func (d *gocryptfsDriver) Start(params *image.DriverParams, containerPid int) error {
+	if containerPid != 0 {
+		// Running in hybrid setuid-fakeroot mode
+		// Need any subcommand to first enter the container's
+		//  user namespace
+		nsenter, err := bin.FindBin("nsenter")
+		if err != nil {
+			return fmt.Errorf("failed to find nsenter: %v", err)
+		}
+		d.cmdPrefix = []string{
+			nsenter,
+			fmt.Sprintf("--user=/proc/%d/ns/user", containerPid),
+			"-F",
+		}
+	}
+	return nil
+}
+
+func (d *gocryptfsDriver) Stop(target string) error {
+	if err := d.gocryptfsFeature.stop(target, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *gocryptfsDriver) GetDriverName() string {
+	return gocryptfsDriverName
+}
+
+func CheckMountInfo(path string, ch chan error) {
+	iter := 0
+	for {
+		ok, err := mountPkg.Mounted(path)
+		if ok {
+			ch <- nil
+			break
+		}
+		if err != nil {
+			ch <- err
+			break
+		}
+		time.Sleep(1 * time.Second)
+		iter++
+		if iter >= 10 {
+			ch <- fmt.Errorf("timeout to check mount info for target path: %s, timeout limit: %d s", path, iter)
+			break
+		}
+	}
+}
```

Comment:
This is similar to code in imagedriver.go, but not as sophisticated.  Sleeping for a second at a time is much too long in my opinion.  Let's make a common function used in both places.